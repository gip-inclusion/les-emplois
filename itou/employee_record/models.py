import contextlib
import json

import django.db.utils
from django.core.exceptions import ValidationError
from django.db import models
from django.db.models.manager import Manager
from django.db.models.query import F, Q, QuerySet
from django.utils import timezone
from rest_framework.authtoken.admin import User

from itou.approvals.models import Approval
from itou.asp.models import EmployerType, PrescriberType, SiaeKind
from itou.job_applications.enums import SenderKind
from itou.siaes.models import Siae, SiaeFinancialAnnex
from itou.users.models import JobSeekerProfile
from itou.utils.validators import validate_siret

from . import constants
from .enums import MovementType, NotificationStatus, NotificationType, Status
from .exceptions import CloningError, DuplicateCloningError, InvalidStatusError


# Validators


def validate_asp_batch_filename(value):
    """
    Simple validation of batch file name
    (ASP backend is picky about it)
    """
    if value and value.startswith("RIAE_FS_") and value.endswith(".json") and len(value) == 27:
        return
    raise ValidationError(f"Le format du nom de fichier ASP est incorrect: {value}")


# Oddly enough, no month param for timedeltas
# => approximate 1 month to 30 days (see base settings)
ARCHIVING_DELTA = timezone.timedelta(days=constants.EMPLOYEE_RECORD_ARCHIVING_DELAY_IN_DAYS)


class ASPExchangeInformation(models.Model):
    ASP_PROCESSING_SUCCESS_CODE = "0000"
    ASP_DUPLICATE_ERROR_CODE = "3436"

    ASP_MOVEMENT_TYPE = None  # Must be specified in descendant classes

    # ASP processing part
    asp_processing_code = models.CharField(max_length=4, verbose_name="Code de traitement ASP", null=True)
    asp_processing_label = models.CharField(max_length=200, verbose_name="Libellé de traitement ASP", null=True)

    # Employee records are sent to ASP in a JSON file,
    # We keep track of the name for processing feedback
    # The format of the file name is EXACTLY: RIAE_FS_AAAAMMJJHHMMSS.json (27 chars)
    asp_batch_file = models.CharField(
        max_length=27,
        verbose_name="Fichier de batch ASP",
        null=True,
        validators=[validate_asp_batch_filename],
    )
    # Line number of the employee record in the batch file
    # Unique pair with `asp_batch_file`
    asp_batch_line_number = models.IntegerField(
        verbose_name="Ligne correspondante dans le fichier batch ASP",
        null=True,
    )

    # Once correctly processed by ASP, the employee record is archived:
    # - it can't be changed anymore
    # - a serialized version of the employee record is stored (as proof, and for API concerns)
    # The API will not use JSON serializers on a regular basis,
    # except for the archive serialization, which occurs once.
    # It will only return a list of this JSON field for archived employee records.
    archived_json = models.JSONField(verbose_name="Archive JSON de la fiche salarié", null=True, blank=True)

    class Meta:
        abstract = True
        constraints = [
            models.UniqueConstraint(
                fields=["asp_batch_file", "asp_batch_line_number"],
                name="unique_%(class)s_asp_batch_file_and_line",
                condition=Q(asp_batch_file__isnull=False),
            )
        ]
        ordering = ["-created_at"]

    def _set_archived_json(self, archive):
        if archive is not None:
            with contextlib.suppress(json.JSONDecodeError):
                archive = json.loads(archive)
        self.archived_json = archive

    def set_asp_batch_information(self, file, line_number, archive):
        self.asp_batch_file = file
        self.asp_batch_line_number = line_number
        self._set_archived_json(archive)

    def set_asp_processing_information(self, code, label, archive):
        self.asp_processing_code = code
        self.asp_processing_label = label
        self._set_archived_json(archive)


class EmployeeRecordQuerySet(models.QuerySet):
    """
    Queryset functions for EmployeeRecord model
    """

    def full_fetch(self):
        """
        Also fetch main employee record related objects:
        - financial annex
        - job application
        - job seeker
        - job seeker profile
        """
        return self.select_related(
            "financial_annex",
            "job_application",
            "job_application__approval",
            "job_application__to_siae",
            "job_application__job_seeker",
            "job_application__job_seeker__birth_country",
            "job_application__job_seeker__birth_place",
            "job_application__job_seeker__jobseeker_profile",
            "job_application__job_seeker__jobseeker_profile__hexa_commune",
        )

    # Search queries
    def for_siae(self, siae):
        return self.filter(
            job_application__to_siae=siae,
            asp_id=F("job_application__to_siae__convention__asp_id"),
        )

    def find_by_batch(self, filename, line_number):
        """
        Fetch a single employee record with ASP batch file input parameters
        """
        return self.filter(asp_batch_file=filename, asp_batch_line_number=line_number)

    def archivable(self):
        """
        Fetch employee records in PROCESSED state for more than EMPLOYEE_RECORD_ARCHIVING_DELAY_IN_DAYS
        """
        return self.filter(status=Status.PROCESSED, processed_at__lte=timezone.now() - ARCHIVING_DELTA)

    def asp_duplicates(self):
        """
        Return REJECTED employee records with error code '3436'.
        These employee records are considered as duplicates by ASP.
        """
        return self.filter(status=Status.REJECTED).filter(asp_processing_code=EmployeeRecord.ASP_DUPLICATE_ERROR_CODE)

    def orphans(self):
        """
        Employee records with an `asp_id` different from their hiring SIAE.
        Could occur when using `siae.move_siae_data` management command.
        """
        return self.exclude(job_application__to_siae__convention__asp_id=F("asp_id"))


class EmployeeRecord(ASPExchangeInformation):
    """
    EmployeeRecord - Fiche salarié (FS for short)

    Holds information needed for JSON exports and processing by ASP
    """

    ERROR_JOB_APPLICATION_MUST_BE_ACCEPTED = "La candidature doit être acceptée"
    ERROR_JOB_APPLICATION_WITHOUT_APPROVAL = "L'embauche n'est pas reliée à un PASS IAE"

    ERROR_JOB_SEEKER_HAS_NO_PROFILE = "Cet utilisateur n'a pas de profil de demandeur d'emploi enregistré"

    ERROR_EMPLOYEE_RECORD_IS_DUPLICATE = "Une fiche salarié pour ce PASS IAE et cette SIAE existe déjà"
    ERROR_EMPLOYEE_RECORD_INVALID_STATE = "La fiche salarié n'est pas dans l'état requis pour cette action"

    ERROR_NO_CONVENTION_AVAILABLE = "La structure actuelle ne dispose d'aucune convention"

    CAN_BE_DISABLED_STATES = [Status.NEW, Status.REJECTED, Status.PROCESSED]

    ASP_MOVEMENT_TYPE = MovementType.CREATION

    created_at = models.DateTimeField(verbose_name="Date de création", default=timezone.now)
    updated_at = models.DateTimeField(verbose_name="Date de modification", auto_now=True)
    processed_at = models.DateTimeField(verbose_name="Date d'intégration", null=True)
    status = models.CharField(max_length=10, verbose_name="Statut", choices=Status.choices, default=Status.NEW)

    # Job application has references on many mandatory parts of the E.R.:
    # - SIAE / asp id
    # - Employee
    # - Approval
    job_application = models.ForeignKey(
        "job_applications.jobapplication",
        on_delete=models.SET_NULL,
        null=True,
        verbose_name="Candidature / embauche",
        related_name="employee_record",
    )

    # Employee records may be linked to a valid financial annex
    # This field can't be automatically filled, the user will be asked
    # to select a valid one manually
    financial_annex = models.ForeignKey(
        SiaeFinancialAnnex, verbose_name="Annexe financière", null=True, on_delete=models.SET_NULL
    )

    # These fields are duplicated to act as constraint fields on DB level
    approval_number = models.CharField(max_length=12, verbose_name="Numéro d'agrément")
    asp_id = models.PositiveIntegerField(verbose_name="Identifiant ASP de la SIAE")

    # If the SIAE is an "antenna",
    # we MUST provide the SIRET of the SIAE linked to the financial annex on ASP side (i.e. "parent/mother" SIAE)
    # NOT the actual SIAE (which can be fake and unrecognized by ASP).
    siret = models.CharField(
        verbose_name="Siret structure mère", max_length=14, validators=[validate_siret], db_index=True
    )

    # When an employee record is rejected with a '3436' error code by ASP, it means that:
    # - all the information transmitted via this employee record is already available on ASP side,
    # - the employee record is not needed by ASP and considered as a duplicate
    # As a result, employee records in `REJECTED` state and with a `3436` error code
    # can safely be considered as `PROCESSED` without side-effects.
    # This field is just a marker / reminder / track record that the status of this object
    # was **forced** to `PROCESSED` (via admin or script) even if originally `REJECTED`.
    # The JSON proof is in this case not available.
    # Forcing a 'PROCESSED' status enables communication for employee record update notifications.
    processed_as_duplicate = models.BooleanField(verbose_name="Déjà intégrée par l'ASP", default=False)

    # Added typing helper: improved type checking for `objects` methods
    objects: EmployeeRecordQuerySet | Manager = EmployeeRecordQuerySet.as_manager()

    class Meta(ASPExchangeInformation.Meta):
        verbose_name = "Fiche salarié"
        verbose_name_plural = "Fiches salarié"
        constraints = ASPExchangeInformation.Meta.constraints + [
            models.UniqueConstraint(
                fields=["asp_id", "approval_number"],
                name="unique_asp_id_approval_number",
                condition=~Q(status=Status.DISABLED),
            )
        ]

    def __str__(self):
        return (
            f"PK:{self.pk} PASS:{self.approval_number} SIRET:{self.siret} JA:{self.job_application} "
            f"JOBSEEKER:{self.job_seeker} STATUS:{self.status} ASP_ID:{self.asp_id}"
        )

    def _clean_job_application(self):
        """
        Check if job application is valid for FS
        """
        if not self.job_application.to_siae.convention:
            raise ValidationError(self.ERROR_NO_CONVENTION_AVAILABLE)

        if not self.job_application.state.is_accepted:
            raise ValidationError(self.ERROR_JOB_APPLICATION_MUST_BE_ACCEPTED)

        if not self.job_application.approval:
            raise ValidationError(self.ERROR_JOB_APPLICATION_WITHOUT_APPROVAL)

    def _clean_job_seeker(self):
        """
        Check if data provided for the job seeker part of the FS is complete / valid
        """
        job_seeker = self.job_application.job_seeker

        # Check if user is "clean"
        job_seeker.clean()

        if not job_seeker.has_jobseeker_profile:
            raise ValidationError(self.ERROR_JOB_SEEKER_HAS_NO_PROFILE)

        # Further validation in the job seeker profile
        # Note that the job seeker profile validation is done
        # via `clean_model` and not `clean` : see comments on `JobSeekerProfile.clean_model`
        job_seeker.jobseeker_profile.clean_model()

    def clean(self):
        # see private methods above
        self._clean_job_application()
        self._clean_job_seeker()

    def _fill_denormalized_fields(self):
        # If the SIAE is an antenna, the SIRET will be rejected by the ASP so we have to use the mother's one
        self.siret = self.siret_from_asp_source(self.job_application.to_siae)
        self.asp_id = self.job_application.to_siae.convention.asp_id
        self.approval_number = self.job_application.approval.number

    # Business methods

    def update_as_ready(self):
        """
        Prepare the employee record for transmission

        Status: NEW | REJECTED => READY
        """
        if self.status not in [Status.NEW, Status.REJECTED]:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        profile = self.job_seeker.jobseeker_profile

        if not profile.hexa_address_filled:
            # Format job seeker address
            profile.update_hexa_address()

        self.job_seeker.last_checked_at = timezone.now()
        self.job_seeker.save(update_fields=["last_checked_at"])

        self.clean()
        # There could be a delay between the moment the object is created
        # and the moment it is completed to be sent to the ASP.
        # In the meantime the ASP ID or the SIRET *can change* (mainly because of weekly ASP import scripts).
        # To prevent some ASP processing errors, we do a refresh on some mutable fields.
        self._fill_denormalized_fields()

        # If we reach this point, the employee record is ready to be serialized
        # and can be sent to ASP
        self.status = Status.READY
        self.save()

    def update_as_sent(self, asp_filename, line_number, archive):
        """
        An employee record is sent to ASP via a JSON file,
        The file name is stored for further feedback processing (also done via a file)

        Status: READY => SENT
        """
        if not self.status == Status.READY:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        self.clean()
        self.status = Status.SENT
        self.set_asp_batch_information(asp_filename, line_number, archive)

        self.save()

    def update_as_rejected(self, code, label, archive):
        """
        Update status after an ASP rejection of the employee record

        Status: SENT => REJECTED
        """
        if not self.status == Status.SENT:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        self.clean()
        self.status = Status.REJECTED
        self.set_asp_processing_information(code, label, archive)

        self.save()

    def update_as_processed(self, code, label, archive):
        if not self.status == Status.SENT:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        self.clean()
        self.status = Status.PROCESSED
        self.processed_at = timezone.now()
        self.set_asp_processing_information(code, label, archive)

        self.save()

    def update_as_disabled(self):
        if not self.can_be_disabled:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        self.status = Status.DISABLED
        self.save()

    def update_as_new(self):
        if self.status != Status.DISABLED:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        self._fill_denormalized_fields()
        # check if FS exists before reactivate
        if (
            EmployeeRecord.objects.exclude(status=Status.DISABLED)
            .filter(asp_id=self.asp_id, approval_number=self.approval_number)
            .exists()
        ):
            raise InvalidStatusError(EmployeeRecord.ERROR_EMPLOYEE_RECORD_IS_DUPLICATE)

        self.status = Status.NEW
        self.save()

    def update_as_archived(self, save=True):
        """
        Can only archive employee record if already PROCESSED
        `save` parameter is for bulk updates in management command
        """
        if self.status != Status.PROCESSED:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        # Check that we are past archiving delay before ... archiving
        if self.processed_at >= timezone.now() - ARCHIVING_DELTA:
            raise InvalidStatusError(self.ERROR_EMPLOYEE_RECORD_INVALID_STATE)

        # Remove proof of processing after delay
        self.status = Status.ARCHIVED
        self.archived_json = None

        if save:
            self.save()
        else:
            # Override .save() update of `updated_at` when using bulk updates
            self.updated_at = timezone.now()

    def update_as_processed_as_duplicate(self, archive):
        """
        Force status to `PROCESSED` if the employee record has been marked
        as duplicate by ASP (error code 3436).

        Can only be done when employee record is:
            - in `REJECTED` state,
            - with a `3436` error code.
        """
        if self.status != Status.REJECTED or self.asp_processing_code != self.ASP_DUPLICATE_ERROR_CODE:
            raise InvalidStatusError(
                f"{self.ERROR_EMPLOYEE_RECORD_INVALID_STATE} ({self.status}, {self.asp_processing_code})"
            )

        self.clean()
        self.status = Status.PROCESSED
        self.processed_at = timezone.now()
        self.processed_as_duplicate = True
        self.set_asp_processing_information(self.ASP_DUPLICATE_ERROR_CODE, "Statut forcé suite à doublon ASP", archive)

        self.save()

    def clone(self):
        """
        Create and return a NEW copy of an employee record, this is useful when orphans are detected.
        If cloning is successful, current employee record is DISABLED (if possible) to avoid conflicts.

        Raises `CloningError` if cloning conditions are not met.
        """
        if not self.pk:
            raise CloningError("This employee record has not been saved yet (no PK).")

        if not self.job_application.to_siae.convention:
            raise CloningError(f"SIAE {self.job_application.to_siae.siret} has no convention")

        # Cleanup clone fields
        er_copy = EmployeeRecord(
            status=Status.NEW,
            job_application=self.job_application,
            approval_number=self.approval_number,
            asp_id=self.job_application.to_siae.convention.asp_id,
            siret=EmployeeRecord.siret_from_asp_source(self.job_application.to_siae),
            asp_processing_label=f"Fiche salarié clonée (pk origine: {self.pk})",
        )

        try:
            er_copy.save()
        except django.db.utils.IntegrityError as ex:
            raise DuplicateCloningError(
                f"The clone is a duplicate of ({er_copy.asp_id=}, {er_copy.approval_number=})"
            ) from ex

        # Disable current object to avoid conflicts
        if self.can_be_disabled:
            self.update_as_disabled()

        return er_copy

    @property
    def can_be_disabled(self):
        return self.status in self.CAN_BE_DISABLED_STATES

    @property
    def job_seeker(self) -> User:
        """
        Shortcut to job application user / job seeker
        """
        return self.job_application.job_seeker if self.job_application else None

    @property
    def job_seeker_profile(self) -> JobSeekerProfile | None:
        """
        Shortcut to job seeker profile
        """
        if self.job_application and hasattr(self.job_application.job_seeker, "jobseeker_profile"):
            return self.job_application.job_seeker.jobseeker_profile

        return None

    @property
    def approval(self) -> Approval | None:
        """
        Shortcut to job application approval
        """
        return self.job_application.approval if self.job_application and self.job_application.approval else None

    @property
    def financial_annex_number(self):
        """
        Shortcut to financial annex number (can be null in early stages of life cycle)
        """
        return self.financial_annex.number if self.financial_annex else None

    @property
    def asp_convention_id(self):
        """
        ASP convention ID (from siae.convention.asp_convention_id)
        """
        if self.job_application and self.job_application.to_siae:
            return self.job_application.to_siae.convention.asp_convention_id

        return None

    @property
    def asp_employer_type(self):
        """
        This is a mapping between itou internal SIAE kinds and ASP ones

        Only needed if profile.is_employed is True

        MUST return None otherwise
        """
        if self.job_seeker_profile and self.job_seeker_profile.is_employed:
            return EmployerType.from_itou_siae_kind(self.job_application.to_siae.kind)
        return None

    # FIXME:
    # This property is currently *never* accepted in production
    # This has to be fixed with ASP (or not)
    # In the meantime the serializer will use a fixed value for this field
    # (see `tmp_prescriber_type` property)
    @property
    def asp_prescriber_type(self):
        """
        This is a mapping between itou internal prescriber kinds and ASP ones
        """

        sender_kind = self.job_application.sender_kind

        if sender_kind == SenderKind.JOB_SEEKER:
            # the job seeker applied directly
            return PrescriberType.SPONTANEOUS_APPLICATION
        elif sender_kind == SenderKind.SIAE_STAFF:
            # an SIAE applied
            # Notify ASP : UNKNOWN code does not work for SIAE
            # FIXME return PrescriberType.UNKNOWN
            return PrescriberType.SPONTANEOUS_APPLICATION

        prescriber_organization = self.job_application.sender_prescriber_organization

        # This workaround is under investigation (systematically fails if UNKNOW is chosen)
        return (
            PrescriberType.from_itou_prescriber_kind(prescriber_organization.kind).value
            if prescriber_organization
            else PrescriberType.AUTHORIZED_PRESCRIBERS
        )

    @property
    def tmp_asp_prescriber_type(self):
        # this is temporary (read above)
        return PrescriberType.SPONTANEOUS_APPLICATION

    @property
    def asp_siae_type(self):
        """
        Mapping between ASP and itou models for SIAE kind ("Mesure")
        """
        return SiaeKind.from_siae_kind(self.job_application.to_siae.kind)

    @property
    def batch_line_number(self):
        """
        This transient field is updated at runtime for JSON serialization.

        It is the batch line number of the employee record.
        """
        if not hasattr(self, "_batch_line_number"):
            self._batch_line_number = 1

        return self._batch_line_number

    @property
    def is_blocking_job_application_cancellation(self):
        """
        Linked job application can't be cancelled if the employee record
        is sent or already processed.
        """
        return self.status in [
            Status.SENT,
            Status.PROCESSED,
        ]

    @property
    def is_orphan(self):
        """Orphan employee records have different stored and actual `asp_id` fields."""
        return self.job_application.to_siae.convention.asp_id != self.asp_id

    @staticmethod
    def siret_from_asp_source(siae):
        """
        Fetch SIRET number of ASP source structure ("mother" SIAE)
        """
        if siae.source != Siae.SOURCE_ASP:
            main_siae = Siae.objects.get(convention=siae.convention, source=Siae.SOURCE_ASP)
            return main_siae.siret

        return siae.siret

    @classmethod
    def from_job_application(cls, job_application):
        """
        Alternative and main FS constructor from a JobApplication object

        If an employee record with given criteria (approval, SIAE/ASP structure)
        already exists, this method returns None

        Defensive:
        - raises exception if job application is not suitable for creation of a new employee record
        - job seeker profile must exist before creating an employee record
        """
        assert job_application

        fs = cls(job_application=job_application)

        fs.clean()

        # Mandatory check, must be done only once
        if (
            EmployeeRecord.objects.exclude(status=Status.DISABLED)
            .filter(
                asp_id=job_application.to_siae.convention.asp_id,
                approval_number=job_application.approval.number,
            )
            .exists()
        ):
            raise ValidationError(EmployeeRecord.ERROR_EMPLOYEE_RECORD_IS_DUPLICATE)

        fs.asp_id = job_application.to_siae.convention.asp_id
        fs.approval_number = job_application.approval.number

        # Fetch correct number if SIAE is an antenna
        fs.siret = EmployeeRecord.siret_from_asp_source(job_application.to_siae)

        return fs


class EmployeeRecordBatch:
    """
    Transient wrapper for a list of employee records.

    Some business validation rules from ASP:
    - no more than 700 employee records per upload
    - serialized JSON file must be 2Mb at most

    This model used by JSON serializer as an header for ASP transmission
    """

    ERROR_BAD_FEEDBACK_FILENAME = "Mauvais nom de fichier de retour ASP"

    # Max number of employee records per upload batch
    MAX_EMPLOYEE_RECORDS = 700

    # Max size of upload file
    MAX_SIZE_BYTES = 2048 * 1024

    # File name format for upload
    REMOTE_PATH_FORMAT = "RIAE_FS_{}.json"

    # Feedback file names end with this string
    FEEDBACK_FILE_SUFFIX = "_FichierRetour"

    def __init__(self, elements):
        if elements and len(elements) > self.MAX_EMPLOYEE_RECORDS:
            raise ValidationError(f"An upload batch can have no more than {self.MAX_EMPLOYEE_RECORDS} elements")

        # id and message fields must be null for upload
        # they may have a value after download
        self.id = None
        self.message = None

        self.elements = elements
        self.upload_filename = self.REMOTE_PATH_FORMAT.format(timezone.now().strftime("%Y%m%d%H%M%S"))

        # add a line number to each FS for JSON serialization
        for idx, er in enumerate(self.elements, start=1):
            er.asp_batch_line_number = idx
            er.asp_processing_code = None
            er.asp_processing_label = None

    def __str__(self):
        return f"FILENAME={self.upload_filename}, NB_RECORDS={len(self.elements)}"

    def __repr__(self):
        # String formating with {field_name=...} forms use __repr__ and not __str__
        return f"{self.upload_filename=}, {len(self.elements)=}"

    @staticmethod
    def feedback_filename(filename):
        """
        Return name of the feedback file
        """
        validate_asp_batch_filename(filename)
        separator = "."
        path, ext = filename.split(separator)
        path += EmployeeRecordBatch.FEEDBACK_FILE_SUFFIX

        return separator.join([path, ext])

    @staticmethod
    def batch_filename_from_feedback(filename):
        """
        Return name of original filename from feedback filename
        """
        separator = "."
        path, ext = filename.split(separator)

        if not path.endswith(EmployeeRecordBatch.FEEDBACK_FILE_SUFFIX):
            raise ValidationError(EmployeeRecordBatch.ERROR_BAD_FEEDBACK_FILENAME)

        # .removesuffix is Python 3.9
        return separator.join([path.removesuffix(EmployeeRecordBatch.FEEDBACK_FILE_SUFFIX), ext])


class EmployeeRecordUpdateNotificationQuerySet(QuerySet):
    def new(self):
        return self.filter(status=NotificationStatus.NEW)

    def sent(self):
        return self.filter(status=NotificationStatus.SENT)

    def processed(self):
        return self.filter(status=NotificationStatus.PROCESSED)

    def rejected(self):
        return self.filter(status=NotificationStatus.REJECTED)

    def find_by_batch(self, filename, line_number):
        return self.filter(asp_batch_file=filename, asp_batch_line_number=line_number)


class EmployeeRecordUpdateNotification(ASPExchangeInformation):
    """
    Notification of PROCESSED employee record updates.

    `NotificationType` details the part of the employee record that is monitored.
    At the moment, only approval updates (start and end dates) are tracked.

    Monitoring of approvals is done via a Postgres trigger (defined in `Approval` app migrations).
    """

    ASP_MOVEMENT_TYPE = MovementType.UPDATE

    created_at = models.DateTimeField(
        verbose_name="Date de création",
        default=timezone.now,
    )
    updated_at = models.DateTimeField(
        verbose_name="Date de modification",
        auto_now=True,
    )
    status = models.CharField(
        verbose_name="Statut",
        max_length=10,
        choices=NotificationStatus.choices,
        default=NotificationStatus.NEW,
    )

    notification_type = models.CharField(
        verbose_name="Type de notification",
        max_length=20,
        choices=NotificationType.choices,
    )

    employee_record = models.ForeignKey(
        EmployeeRecord,
        related_name="update_notifications",
        verbose_name="Fiche salarié",
        on_delete=models.CASCADE,
    )

    objects = EmployeeRecordUpdateNotificationQuerySet.as_manager()

    class Meta(ASPExchangeInformation.Meta):
        verbose_name = "Notification de changement de la fiche salarié"
        verbose_name_plural = "Notifications de changement de la fiche salarié"

    def __repr__(self):
        return f"<{type(self).__name__} pk={self.pk}>"

    def update_as_sent(self, filename, line_number, archive):
        if self.status not in [NotificationStatus.NEW, NotificationStatus.REJECTED]:
            raise ValidationError(f"Invalid status to update as SENT (currently: {self.status})")

        self.status = NotificationStatus.SENT
        self.set_asp_batch_information(filename, line_number, archive)

        self.save()

    def update_as_rejected(self, code, label, archive):
        if not self.status == Status.SENT:
            raise ValidationError(f"Invalid status to update as REJECTED (currently: {self.status})")

        self.status = Status.REJECTED
        self.set_asp_processing_information(code, label, archive)

        self.save()

    def update_as_processed(self, code, label, archive):
        if not self.status == Status.SENT:
            raise ValidationError(f"Invalid status to update as PROCESSED (currently: {self.status})")

        self.status = Status.PROCESSED
        self.set_asp_processing_information(code, label, archive)

        self.save()
