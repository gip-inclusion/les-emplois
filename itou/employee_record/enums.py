from django.db import models


class Status(models.TextChoices):
    """
    Status of the employee record

    Self-explanatory on the meaning, however:
    - an employee record can be modified until it is in the PROCESSED state
    - after that, the object is "archived" and can't be used for further interaction
    """

    NEW = "NEW", "Nouvelle"
    READY = "READY", "Complétée"
    SENT = "SENT", "Envoyée"
    REJECTED = "REJECTED", "En erreur"
    PROCESSED = "PROCESSED", "Intégrée"
    ARCHIVED = "ARCHIVED", "Archivée"


class NotificationStatus(models.TextChoices):
    NEW = "NEW", "Nouvelle"
    SENT = "SENT", "Envoyée"
    PROCESSED = "PROCESSED", "Intégrée"
    REJECTED = "REJECTED", "En erreur"


class NotificationType(models.TextChoices):
    """
    Add a new type for each new employee record update notification.
    """

    APPROVAL = "APPROVAL", "Modification du PASS IAE"
    JOB_SEEKER = "JOB_SEEKER", "Modification de l'employé"


class MovementType(models.TextChoices):
    CREATION = "C", "Création"
    UPDATE = "M", "Modification"
