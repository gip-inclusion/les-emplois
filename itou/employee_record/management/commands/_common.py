from io import BytesIO
from os import path
from typing import Optional

import pysftp
from django.conf import settings
from django.core.management.base import BaseCommand
from django.utils import timezone
from rest_framework.renderers import JSONRenderer


# Global SFTP connection options

connection_options = None

if settings.ASP_FS_KNOWN_HOSTS and path.exists(settings.ASP_FS_KNOWN_HOSTS):
    connection_options = pysftp.CnOpts(knownhosts=settings.ASP_FS_KNOWN_HOSTS)


class EmployeeRecordTransferCommand(BaseCommand):
    def get_sftp_connection(self) -> pysftp.Connection:
        """
        Get a new SFTP connection to remote server.
        """
        return pysftp.Connection(
            host=settings.ASP_FS_SFTP_HOST,
            port=int(settings.ASP_FS_SFTP_PORT),
            username=settings.ASP_FS_SFTP_USER,
            private_key=settings.ASP_FS_SFTP_PRIVATE_KEY_PATH,
            cnopts=connection_options,
        )

    def upload_json_file(self, json_data, conn: pysftp.Connection, dry_run=False) -> Optional[str]:
        """
        Upload `json_data` (as byte array) to given SFTP connection `conn`.
        Returns uploaded filename if ok, `None` otherwise.
        """
        # JSONRenderer produces *byte array* not strings
        json_bytes = JSONRenderer().render(json_data)
        now = timezone.now()

        # Using FileIO objects allows to use them as files
        # Cool side effect: no temporary file needed
        json_stream = BytesIO(json_bytes)
        remote_path = f"RIAE_FS_{now:%Y%m%d%H%M%S}.json"

        if dry_run:
            self.stdout.write(f"DRY-RUN: (not) sending '{remote_path}' ({len(json_bytes)} bytes)")
            self.stdout.write(f"Content: \n{json_bytes}")

            return remote_path

        # There are specific folders for upload and download on the SFTP server
        with conn.cd(settings.ASP_FS_REMOTE_UPLOAD_DIR):
            # After creating a FileIO object, internal pointer is at the end of the buffer
            # It must be set back to 0 (rewind) otherwise an empty file is sent
            json_stream.seek(0)

            # ASP SFTP server does not return a proper list of transmitted files
            # Whether it's a bug or a paranoid security parameter
            # we must assert that there is no verification of the remote file existence
            # This is the meaning of `confirm=False`
            try:
                conn.putfo(json_stream, remote_path, file_size=len(json_bytes), confirm=False)

                self.stdout.write(f"Successfully uploaded: {remote_path}")
            except Exception as ex:
                self.stdout.write(f"Could not upload file: {remote_path}, reason: {ex}")
                return

        return remote_path

    def download_json_file(self):
        # FIXME
        pass
