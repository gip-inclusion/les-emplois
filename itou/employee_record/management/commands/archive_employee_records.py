from django.core.management.base import BaseCommand
from django.db import transaction

from itou.employee_record.models import EmployeeRecord


class Command(BaseCommand):
    def add_arguments(self, parser):
        super().add_arguments(parser)

        parser.add_argument("--wet-run", action="store_true")

    @transaction.atomic()
    def handle(self, wet_run=False, **options):
        self.stdout.write("Start archiving employee records")

        archivable = (
            EmployeeRecord.objects.archivable()
            .order_by("job_application__approval__end_at")
            .select_related("job_application__approval")
        )
        self.stdout.write(f"Found {len(archivable)} archivable employee record(s)")

        archived_employee_records = []
        # A bulk update will increase performance if there are a lot of employee records to update.
        # However, if there is no performance issue, it is preferable to keep the archiving
        # and validation logic in the model (update_as_archived).
        # Update: let's bulk, with a batch size of 100 records
        for employee_record in archivable:
            try:
                # Do not trigger a save() call on the object
                self.stdout.write(f"Archiving {employee_record.pk=}")
                employee_record.update_as_archived(save=False)
            except Exception as ex:
                self.stdout.write(f"Can't archive {employee_record.pk=} {ex=}")
            else:
                archived_employee_records.append(employee_record)

        self.stdout.write(f"{len(archived_employee_records)}/{len(archivable)} employee record(s) can be archived")

        if wet_run:
            updated = EmployeeRecord.objects.bulk_update(
                archived_employee_records,
                ["status", "updated_at", "archived_json"],
                batch_size=100,
            )
            self.stdout.write(
                f"{updated}/{len(archived_employee_records)}/{len(archivable)} employee record(s) were archived"
            )
