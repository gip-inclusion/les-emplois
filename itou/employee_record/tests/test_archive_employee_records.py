import io

import pytest
from django.core.management import call_command

from .. import factories
from ..enums import Status
from ..management.commands import archive_employee_records
from ..models import EmployeeRecord


@pytest.fixture(name="command")
def command_fixture():
    return archive_employee_records.Command(stdout=io.StringIO(), stderr=io.StringIO())


def test_management_command_default_run(command, snapshot):
    employee_record = factories.EmployeeRecordFactory(pk=42, archivable=True)
    assert list(EmployeeRecord.objects.archivable()) == [employee_record]

    command.handle(wet_run=False)
    employee_record.refresh_from_db()

    assert employee_record.status != Status.ARCHIVED
    assert command.stdout.getvalue() == snapshot


def test_management_command_wet_run(command, snapshot):
    employee_record = factories.EmployeeRecordFactory(pk=42, archivable=True)

    command.handle(wet_run=True)
    employee_record.refresh_from_db()

    assert employee_record.status == Status.ARCHIVED
    assert employee_record.archived_json is None
    assert command.stdout.getvalue() == snapshot


def test_management_command_name(faker):
    call_command("archive_employee_records")
