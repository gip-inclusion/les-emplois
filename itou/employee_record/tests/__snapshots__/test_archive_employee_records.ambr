# serializer version: 1
# name: test_management_command_default_run
  '''
  Start archiving employee records
  Found 1 archivable employee record(s)
  Archiving employee_record.pk=42
  1/1 employee record(s) can be archived
  
  '''
# ---
# name: test_management_command_wet_run
  '''
  Start archiving employee records
  Found 1 archivable employee record(s)
  Archiving employee_record.pk=42
  1/1 employee record(s) can be archived
  1/1/1 employee record(s) were archived
  
  '''
# ---
